autonomous mavic

joachimhgg opened the pull request:
This PR aims to develop autonomous drones patroling over the forest, detecting and dropping water on wildfires.

An issue appeared with the method `ComputeMonoplotting()`. It aims to compute the world coordinates from the image coordinates of the fire. It gets the image coordinates of the fire and return its world coordinates. If its works in some situations, the robot goes sometimes away the fire instead of going in, or not close enough; due to wrong given world coordinates. The method is based on http://sar.kangwon.ac.kr/etc/rs_note/rsnote/cp9/cp9-6.htm taken from https://stackoverflow.com/questions/68195056/finding-the-real-world-coordinates-of-an-object-from-a-camera 

To overtake this issue, I used an alternative solution which can be found into the code.

joachimhgg committed:
first commit

omichel commented on projects/forest_firefighters/controllers/fire/fire.py:
```diff
@@ -229,40 +245,44 @@ def checkExtinction(self, tree):  # check and extinct the fire if there is water
                         tree.stopFire()
 
     def run(self):
+        flag_start_sim = False
```
I would rename this variable as `start_fire_now`.

omichel commented on projects/forest_firefighters/protos/Smoke.proto:
```diff
@@ -0,0 +1,46 @@
+          texCoordIndex [
+            2 0 1 3 -1
+          ]
+       }
```
```suggestion
        }
```

omichel commented on projects/forest_firefighters/worlds/.forest_firefighters.wbproj:
```diff
@@ -5,14 +5,8 @@ sceneTreePerspectives: 000000ff0000000100000002000000c0000000fc01000000060100000
 orthographicViewHeight: 24.5163
 textFiles: -1
-robotWindow: robot
-globalOptionalRendering: CoordinateSystem
+globalOptionalRendering: CoordinateSystem::RadarFrustums::CameraFrustums::ContactPoints
 consoles: Console:All:All
-renderingDevicePerspectives: Mavic 2 PRO:camera;1;0.6625;0;0
-renderingDevicePerspectives: Spot:left flank camera;0;0.790741;0;0
-renderingDevicePerspectives: Spot:left head camera;0;0.790741;0;0
-renderingDevicePerspectives: Spot:rear camera;0;0.790741;0;0
-renderingDevicePerspectives: Spot:right flank camera;0;0.790741;0;0
-renderingDevicePerspectives: Spot:right head camera;0;0.790741;0;0
```
It doesn't seem good to commit these changes.

omichel commented on projects/forest_firefighters/worlds/forest_firefighters.wbt:
```diff
@@ -223,17 +223,87 @@ SquareManhole {
   translation 4.36 4.12 16.79
   rotation 0 0 1 0
-  controller "mavic"
+  controller "autonomous_mavic_v4"
```
I would remove the `_v4` suffix in the final version.

omichel commented on projects/forest_firefighters/controllers/autonomous_mavic_v4/autonomous_mavic_v4.py:
```diff
@@ -0,0 +1,347 @@
+        verbose (bool): whether to print status messages or not
+        """
+        resolution = self.camera.getWidth(), self.camera.getHeight()
+        roll = self.curr_pos[3]
+        pitch = self.curr_pos[4]
+        yaw = self.curr_pos[5]
```
This is the roll, pitch, yaw of the IMU, the ones of the camera may be different.

omichel commented on projects/forest_firefighters/controllers/autonomous_mavic_v4/autonomous_mavic_v4.py:
```diff
@@ -0,0 +1,347 @@
+        Z_obj = 30
+
+        #intrinsic properties of the camera
+        cx, cy, cz = self.curr_pos[0:3] #principal point in the image ~= real world coordinates of the camera
```
This is the position of the GPS, not necessarily the one of the camera.

omichel commented on projects/forest_firefighters/controllers/autonomous_mavic_v4/autonomous_mavic_v4.py:
```diff
@@ -0,0 +1,347 @@
+
+        fieldOfView = self.camera.getFov() #0.78398rad FOV = 2*arctan(x/2f)
+        diag_img = np.sqrt(resolution[0]**2 + resolution[1]**2)
+        f = diag_img/(2*np.tan(fieldOfView/2)) #f = focal length
```
That seems wrong, the focal length doesn't depend on the image resolution in pixels.

joachimhgg replied:
That's true. The focal length depends of the diagonal of the film or sensor, but it should be in m and not in pixel. Is it possible to get the size of a pixel in m or the diagonal size of the sensor?

omichel commented on projects/forest_firefighters/controllers/autonomous_mavic_v4/autonomous_mavic_v4.py:
```diff
@@ -0,0 +1,347 @@
+
+        #intrinsic properties of the camera
+        cx, cy, cz = self.curr_pos[0:3] #principal point in the image ~= real world coordinates of the camera
+        f = self.camera.getFocalLength() #0, not working
```
Yes, this is documented. It returns 0 if the camera has no Focus node.

omichel reviewed (commented):
I found several possible sources for the error, see my comments.

joachimhgg committed:
Update projects/forest_firefighters/protos/Smoke.proto

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

joachimhgg committed:
add naiveapproach + fixes

joachimhgg committed:
Merge branch 'develop-autonomous-mavic' of https://github.com/cyberbotics/webots-projects into develop-autonomous-mavic

joachimhgg committed:
fix angles

joachimhgg committed:
add readme + clean code

joachimhgg committed:
Update README.md

joachimhgg committed:
Update README.md

omichel commented on projects/forest_firefighters/README.md:
```diff
@@ -77,3 +77,11 @@ The keyboard can be used to move the robot and drop water on the forest fire.
 ## Spot Demo Controller
 The Spot controller is a basic controller that just moves the legs of the robot. The keyboard can also be used to throw water on the forest fire 
+
+## Autonomous mavic Controller
```
```suggestion
## Autonomous Mavic Controller

```

omichel commented on projects/forest_firefighters/README.md:
```diff
@@ -77,3 +77,11 @@ The keyboard can be used to move the robot and drop water on the forest fire.
+
+## Autonomous mavic Controller
+
+The autonomous mavic moves between chosen coordinates above a forest in order to patrol and detect a fire. The altitude and world coordinates can be chosen through the controller args. Each time a fire is detected, the drone goes above the fire and drop a quantity of water to stop the fire. The project is using OpenCV to detect the smoke color range, and a naive approach to move above the fire (m ...
```
We usually write one sentence per line.
```suggestion
The autonomous Mavic drone moves between chosen coordinates above a forest in order to patrol and detect a fire.
The altitude and world coordinates can be chosen through the controller args.
Each time a fire is detected, the drone goes above the fire and drop a quantity of water to stop the fire.
The project is using OpenCV to detect the smoke color range, and a naive approach to move above the fire (moves until the fire is in the center of the image).
An alternative approach could be to use monoplotting to compute the world coordinates from the image coordinates of the fire (for example based on [this method](http://sar.kangwon.ac.kr/etc/rs_note/rsnote/cp9/cp9-6.htm)).  

```

omichel commented on projects/forest_firefighters/README.md:
```diff
@@ -77,3 +77,11 @@ The keyboard can be used to move the robot and drop water on the forest fire.
+
+The autonomous mavic moves between chosen coordinates above a forest in order to patrol and detect a fire. The altitude and world coordinates can be chosen through the controller args. Each time a fire is detected, the drone goes above the fire and drop a quantity of water to stop the fire. The project is using OpenCV to detect the smoke color range, and a naive approach to move above the fire (m ...
+
+It is also possible to add multiple drones to increase chances of detecting a fire. Please note that the fire is starting only when the drone have reached an altitude of 40m.  
```
```suggestion
It is also possible to add multiple drones to increase chances of detecting a fire.
Please note that the fire is starting only when the drone has reached an altitude of 40m.
```

omichel commented on projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py:
```diff
@@ -0,0 +1,285 @@
+
+    MAX_YAW_DISTURBANCE = 0.4
+    MAX_PITCH_DISTURBANCE = -1
+    target_precison = 0.5 #precision between the target position and the robot position in meters
```
Please check PEP8 compliance. See https://github.com/cyberbotics/webots/wiki/Python-Coding-Style
```suggestion
    target_precision = 0.5  # precision between the target position and the robot position in meters
```

omichel commented on projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py:
```diff
@@ -0,0 +1,285 @@
+
+        self.timeStep = int(self.getBasicTimeStep())
+        
+        self.water_to_drop = 0
```
This is not consistent:
```suggestion
        self.time_step = int(self.getBasicTimeStep())
        
        self.water_to_drop = 0
```

omichel commented on projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py:
```diff
@@ -0,0 +1,285 @@
+            motor.setVelocity(1)
+
+        self.curr_pos = 6*[0] # X,Y,Z, yaw, pitch, roll
+        self.target_pos = [0,0,0]
```
Please do not use abbreviations, they are often confusing as here:
```suggestion
        self.current_pose = 6*[0] # X,Y,Z, yaw, pitch, roll
        self.target_position = [0,0,0]
```

omichel commented on projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py:
```diff
@@ -0,0 +1,285 @@
+            print("remaning angle: {:.4f}, remaning distance: {:.4f}".format(angle_left, distance_left))
+        return yaw_disturbance, pitch_disturbance
+
+    def NaiveApproach(self, verbose=True):
```
Consistency:
```suggestion
    def naive_approach(self, verbose=True):
```

omichel commented on projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py:
```diff
@@ -0,0 +1,285 @@
+
+    def NaiveApproach(self, verbose=True):
+        """
+        Naive approch to move the robot above the fire. Closed loop to move the robot towards to the fire by step until to be above it.
```
```suggestion
        Naive approach to move the robot above the fire. Closed loop to move the robot towards to the fire step-by-step until it reaches the fire.
```

omichel commented on projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py:
```diff
@@ -0,0 +1,285 @@
+
+        return yaw_disturbance, pitch_disturbance
+
+    def FireDetection(self, verbose=True):
```
```suggestion
    def fire_detection(self, verbose=True):
```

omichel commented on projects/forest_firefighters/worlds/forest_firefighters.wbt:
```diff
@@ -223,19 +223,41 @@ SquareManhole {
+  controller "autonomous_mavic"
+  controllerArgs [
+    "--target_altitude=45"
+    "--patrol_coord=11 11, 21 11, 21 21, 11 21"
```
```suggestion
    "--patrol_coords=11 11, 21 11, 21 21, 11 21"
```

omichel reviewed (changes requested):
I believe we should not remove the spot robot from the simulation.
Also, there are only two drones, can't we have 4?

joachimhgg committed:
Update projects/forest_firefighters/README.md

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

joachimhgg committed:
Update projects/forest_firefighters/README.md

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

joachimhgg committed:
Update projects/forest_firefighters/README.md

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

joachimhgg committed:
Update projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

joachimhgg committed:
Update projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

joachimhgg committed:
Update projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py

Co-authored-by: Olivier Michel <Olivier.Michel@cyberbotics.com>

joachimhgg committed:
consistency + PEP8 compliance fixes

joachimhgg committed:
Update README.md

omichel commented:
Thank you, the code looks better. What about my original comments:

> I believe we should not remove the spot robot from the simulation.
> Also, there are only two drones, can't we have 4?

joachimhgg committed:
add Spot robot

joachimhgg committed:
Merge branch 'develop-autonomous-mavic' of https://github.com/cyberbotics/webots-projects into develop-autonomous-mavic

joachimhgg committed:
stop spot controller after some time

omichel commented:
Thank you. I believe we should improve this simulation with the following:
- [x] Don't display the camera overlays.
- [x] Fix the depth fighting problem with the Mavic platforms (move them slightly higher).
- [x] Fix the 'D' key control for the Spot robot.
- [x] Restart random fires repeatedly so that we can see the Mavic robots in action more than once.
- [x] Make a small movie showing the Mavic robots fighting fires.

joachimhgg committed:
improve simulation

joachimhgg committed:
Update README.md

joachimhgg commented:
> Thank you. I believe we should improve this simulation with the following:
> 
>     * [x]  Don't display the camera overlays.
> 
>     * [x]  Fix the depth fighting problem with the Mavic platforms (move them slightly higher).
> 
>     * [x]  Fix the 'D' key control for the Spot robot.
> 
>     * [x]  Restart random fires repeatedly so that we can see the Mavic robots in action more than once.
> 
>     * [x]  Make a small movie showing the Mavic robots fighting fires.

I fixed the different issues and added a video to the Readme

omichel committed:
Update README.md

omichel reviewed (dismissed):
All good. Thank you!

omichel reviewed (changes requested):
It seems the robot window for controlling the wind is broken on this branch...

joachimhgg committed:
display wind controller

joachimhgg committed:
Merge branch 'develop-autonomous-mavic' of https://github.com/cyberbotics/webots-projects into develop-autonomous-mavic

joachimhgg committed:
Update README.md

omichel committed:
Fixed robot window on Windows

omichel reviewed (approved):
Looks good to me now.
Thank you.
